### ISSUE TITLE
InferenceSession fails when fp16 model is loaded with CPUExecutionProvider

### ISSUE THREAD

**Issue by SergeySandler** [microsoft/onnxruntime#15494]
Converting ResNet50 model from _TensorFlow_ into _ONNX_ with tf2onnx.convert produces a model that can be used on CPU as well as on GPU.

Further, converting weights into _FP16_ with onnxconverter_common.float16.convert_float_to_float16() produces a new model that can be loaded with _DmlExecutionProvider_, but fails with _CPUExecutionProvider_. It fails in a wrong way, with _The kernel appears to have died. It will restart automatically._

The model is attached (as a multiple disk archive due to maximum size restriction, rename ResNet50.fp16.z01.zip as ResNet50.fp16.z01). To reproduce, run

```
import onnxruntime as ort
model = ort.InferenceSession("ResNet50.fp16.onnx", 
              providers = ['CPUExecutionProvider'], sess_options = ort.SessionOptions(), provider_options = None)

```
_pip show onnxruntime_ reports _Version: 1.14.1_

[ResNet50.fp16.zip](https://github.com/microsoft/onnxruntime/files/11218742/ResNet50.fp16.zip)
[ResNet50.fp16.z01.zip](https://github.com/microsoft/onnxruntime/files/11218745/ResNet50.fp16.z01.zip)

**Comment by SergeySandler** [comment]
@chenfucn it is x64, Intel i7. The problem happens in both, natively in Windows and in a Docker container.

**Comment by chenfucn** [comment]
The attached model runs without issue with the C++ binary. Need time to investigate further as I don't usually work with python.

**Comment by chenfucn** [comment]
The converted model can't pass onnx checker:

```
import onnx
onnx_model = onnx.load("ResNet50.fp16.onnx")
onnx.checker.check_model(onnx_model)
```

And it produces error message: 
```
onnx.onnx_cpp2py_export.checker.ValidationError: Nodes in a graph must be topologically sorted, however input 'graph_input_cast_0' of node:
name: StatefulPartitionedCall/resnet50/conv1_conv/Conv2D__6 OpType: Transpose
 is not output of any previous nodes.
```

 Please file an issue with the converter you're using. Thanks!

**Comment by SergeySandler** [comment]
@chenfucn the same issue is reproducible with C++, Session::Session() constructor causes multiple
`Exception thrown at 0x00007FFCB3FFEC7A (onnxruntime.dll) in inferenceONNX.exe: 0xC0000005: Access violation reading location 0x0000000000000020.`
until program terminates, presumably with the stack overflow. As in Python, it happens with _CPUExecutionProvider_ only. 

What version of _onnxruntime_ were you testing with? It is not 1.14.1, is it?

I can reproduce the error produced by _onnx.checker.check_model(onnx_model)_. It is desirable ort.InferenceSession() fails gracefully in any case, not miserably with the application terminated. Do you agree?

**Comment by chenfucn** [comment]
Yeah I am using onnxruntime repo main branch. And I just tried it on branch rel-1.14.1, it does get seg fault.  I guess someone in our team fixed the problem since the last release.

@yuslepukhin Would you know who might be the white knight? The problem was in graph.cc Node::TryGetFunctionProto(). 

```
    } else if (op_->HasFunction()) {
      onnx_function_proto = *(op_->GetFunction());
      return true;
    }
```
the thing is, in rel-1.14.1, even after op_->HasFunction() returned true, op->GetFunction() returned nullptr.  does not happen in the head of main.
